### PR TITLE
Logproxy in router

### PR DESCRIPTION
**Pull request by debebantur**

### CODE REVIEW

File: pkg/meta/meta.go [hunk 1]
```diff
@@ -110,8 +111,16 @@ func processCreate(ctx context.Context, astmt spqrparser.Statement, mngr EntityM
 	}
 }
 
-func Proc(ctx context.Context, tstmt spqrparser.Statement, mgr EntityMgr, ci connectiterator.ConnectIterator, cli *clientinteractor.PSQLInteractor) error {
+func Proc(ctx context.Context, tstmt spqrparser.Statement, mgr EntityMgr, ci connectiterator.ConnectIterator, cli *clientinteractor.PSQLInteractor, writer workloadlog.WorkloadLogIface) error {
 	switch stmt := tstmt.(type) {
+	//Case new message type
+	// add flag/channel/ to entityMgr or create new
+	case *spqrparser.TraceStmt:
+		cli.CompleteMsg(7)
```
Comment (reshke) [inline pkg/meta/meta.go:119]: what does 7 mean?

File: pkg/workloadlog/workloadlog.go [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+	go serv(wl.ch, wl.ctx)
+	spqrlog.Zero.Error().Msg("started gorutine")
+	return nil
+}
+
+func (wl *WorkloadLogger) IsLogging() bool {
+	return !(wl.Mode == None)
+}
+
+func (wl *WorkloadLogger) WriteLog(msg pgproto3.FrontendMessage, client string) {
```
Comment (reshke) [inline pkg/workloadlog/workloadlog.go:68]: Maybe `RecordWorkload`?

File: pkg/workloadlog/workloadlog.go [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+				spqrlog.Zero.Err(err).Msg("")
+			}
+			return nil
+		case tm := <-ch:
+			byt, err := encodeMessage(tm)
+			if err != nil {
+				spqrlog.Zero.Err(err).Msg("")
+			}
+			interData = append(interData, byt...)
+			if len(interData) > 1000000 {
```
Comment (reshke) [inline pkg/workloadlog/workloadlog.go:104]: `if len(...) > VARIABLE_SIZE` where VARIABLE_SIZE is configurable via config or other ways.

File: pkg/workloadlog/workloadlog.go [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+				}
+				interData = []byte{}
+			}
+		}
+	}
+}
+
+func flush(interceptedData []byte) error {
+	spqrlog.Zero.Error().Msg("saving file logs")
+	f, err := os.OpenFile("mylogs.txt", os.O_APPEND|os.O_WRONLY|os.O_CREATE, 0600)
```
Comment (reshke) [inline pkg/workloadlog/workloadlog.go:117]: os.OpenFile(logsPath, ...) where logsPath is configurable

File: coordinator/provider/coordinator.go [hunk 5]
```diff
@@ -1078,7 +1078,7 @@ func (qc *qdbCoordinator) ProcClient(ctx context.Context, nconn net.Conn) error
 				Type("type", tstmt).
 				Msg("parsed statement is")
 
-			if err := meta.Proc(ctx, tstmt, qc, ci, cli); err != nil {
+			if err := meta.Proc(ctx, tstmt, qc, ci, cli, nil); err != nil {
```
Comment (reshke) [inline coordinator/provider/coordinator.go:1081]: So `START TRACE` requests to coordinator will end up with panic?

File: pkg/workloadlog/workloadlog.go [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+		case tm := <-wl.messageQueue:
+			byt, err := encodeMessage(tm)
+			if err != nil {
+				spqrlog.Zero.Err(err).Msg("")
+			}
+			interData = append(interData, byt...)
+			if len(interData) > wl.batchSize {
+				err = flush(interData, wl.logFile)
+				if err != nil {
+					spqrlog.Zero.Err(err).Msg("")
```
Comment (reshke) [inline pkg/workloadlog/workloadlog.go:135]: Maybe more meaningful msg here?

File: yacc/console/gram.y [hunk 7]
```diff
@@ -362,10 +361,14 @@ trace_stmt:
 	{
 		$$ = &TraceStmt{All: true}
 	} | 
-	START TRACE CLIENT any_id {
+	START TRACE CLIENTID any_id {
```
Comment (reshke) [inline yacc/console/gram.y:364]: why CLIENTID?

Comment (reshke) [reply yacc/console/gram.y:364]: All other cmds uses CLIENT, like KILL CLIENT or SHOW ... WHERE CLIEND

File: yacc/console/gram.y [hunk 8]
```diff
@@ -362,10 +361,14 @@ trace_stmt:
...
 		$$ = &TraceStmt {
 			ClientID: $4,
 		}
+	} |
+	STOP TRACE MESSAGES
+	{
+		$$ = &TraceStmt{Stop: true}
```
Comment (reshke) [inline yacc/console/gram.y:371]: We can define separate structure for this case

File: yacc/console/gram.y (hunk 7 above)
Comment (debebantur) [reply yacc/console/gram.y:364]: It conflicts with other commands. When CLIENT is in reserved_keyword.go commands like kill client do not work, and when CLIENT is not in the list START TRACE CLIENT does not work

Comment (debebantur) [reply yacc/console/gram.y:364]: If you have any idea how to fix this it would be great

File: yacc/console/lx_test.go [hunk 9]
```diff
@@ -24,7 +24,7 @@ func TestSimpleLex(t *testing.T) {
 		},
 		{
 			query: "kill client 0xc00030f520;",
-			exp:   []int{spqrparser.KILL, spqrparser.IDENT, spqrparser.IDENT},
+			exp:   []int{spqrparser.KILL, spqrparser.CLIENT, spqrparser.IDENT},
```
Comment (reshke) [inline yacc/console/lx_test.go:27]: maybe we can add test on new token types(introduces in this pr?)

File: yacc/console/gram.y [hunk 10]
```diff
@@ -54,6 +54,7 @@ func randomHex(n int) (string, error) {
 	listen                 *Listen
 
 	trace                  *TraceStmt
+	stoptrace                  *StopTraceStmt
```
Comment (reshke) [inline yacc/console/gram.y:57]: fmt

File: yacc/console/ast.go [hunk 11]
```diff
@@ -50,12 +50,18 @@ type CreateStmt interface {
 }
 
 type TraceStmt struct {
-	ClientID string
-	All      bool
+	Client string
+	All    bool
+	Stop   bool
```
Comment (reshke) [inline yacc/console/ast.go:55]: so we can drop `Stop` field?

File: yacc/console/lx_test.go (hunk 9 above)
Comment (debebantur) [reply yacc/console/lx_test.go:27]: you mean so, that expected will be []int{spqrparser.KILL, spqrparser.IDENT, spqrparser.IDENT} ?

Comment (debebantur) [reply yacc/console/lx_test.go:27]: > maybe we can add test on new token types(introduces in this pr?)

you mean so, that expected will be []int{spqrparser.KILL, spqrparser.IDENT, spqrparser.IDENT} ?

**Review by reshke — APPROVED** [review]

**Pull request merged**